atomicapp/providers/lib/kubeshift/kubebase.py

```diff
@@ -0,0 +1,331 @@
+            raise KubeConnectionError(msg)
+        except requests.exceptions.ReadTimeout:
+            msg = "Timeout when reading from %s" % url
+            raise KubeConnectionError(msg)
+        except requests.exceptions.ConnectionError:
+            msg = "Refused connection to %s" % url
+            raise KubeConnectionError(msg)
+        except SSLError:
+            raise KubeConnectionError("SSL/TLS ERROR: invalid certificate")
+        except ValueError:
+            return_data = None
+
+        # 200 = OK
+        # 201 = PENDING
+        # EVERYTHING ELSE == FAIL
+        if status_code is not 200 and status_code is not 201:
+            raise KubeConnectionError("Unable to complete request: Status: %s, Error: %s"
+                                      % (status_code, return_data))
+        return return_data
+
+    def websocket_request(self, url, outfile=None):
+        '''
+        Due to the requests library not supporting SPDY, websocket(s) are required
+        to communicate to the API.
+
+        Args:
+            url (str): URL of the API
+            outfile (str): path of the outfile/data.
+        '''
+        url = 'wss://' + url.split('://', 1)[-1]
+        logger.debug('Converted http to wss url: {}'.format(url))
+        results = []
+
+        ws = websocket.WebSocketApp(
+            url,
+            on_message=lambda ws, message: self._handle_exec_reply(ws, message, results, outfile))
+
+        ws.run_forever(sslopt={
+            'ca_certs': self.cert_ca if self.cert_ca is not None else ssl.CERT_NONE,
+            'cert_reqs': ssl.CERT_REQUIRED if self.skip_tls else ssl.CERT_NONE})
+
+        # If an outfile was not provided, return the results in its entirety
+        if not outfile:
+            return ''.join(results)
+
+    def get_resources(self, url):
+        '''
+        Get the resources available to the API. This is a list of all available
+        API calls that can be made to the API.
+        '''
+        data = self.request("get", url)
+        resources = data["resources"]
+        resources = [res['name'] for res in resources]
+        return resources
+
+    def test_connection(self, url):
+        self.api.request("get", url)
+        logger.debug("Connection successfully tested")
+
+    @staticmethod
+    def cert_file(data):
+        '''
+        Some certificate .kube/config components are required to be a filename.
+
+        Returns either the filename or a tmp location of said data in a file.
+        All certificates used with Kubernetes are base64 encoded and thus need to be decoded
+        '''
+
+        # If it starts with /, we assume it's a filename so we just return that.
+        if data.startswith('/'):
+            return data
+
+        # If it's data, we assume it's a base64 certificate and we decode and write to a tmp file
+        else:
+            with tempfile.NamedTemporaryFile(delete=False) as f:
+                f.write(base64.b64decode(data))
```

Comment:
Thanks for this @kadel :) I've updated the PR to reflect what parts should be decoded and what shouldn't.

I'll have to add the other `-data` keys in a separate PR